packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -287,6 +288,7 @@ export class BlobManager {
     }
 
     public async createBlob(blob: ArrayBufferLike): Promise<IFluidHandle<ArrayBufferLike>> {
```

Comment:
we should add the concept of dirty to the blob manager itself, and then hook into that from the container runtime, not try to have the runtime independently manage dirty state around blobs. this should also be more unit testable

Reply:
I'd disagree. Dirty means - there are changes that were already applied and reflected in data model (and thus very likely - on the screen) that were not saved. Pending blob upload does not satisfy that criteria - there are no changes to local data model until upload fully commits to data model.
Exception is offline blob creation flow.

Reply:
Litmus test for me - are changes on the screen only for local client? If they are not, or if they are reflected on remote machine, then document is not dirty. If app has a layer of "lies" on top of local FF state, then they need to incorporate such lies into their calculation of dirty state tracking - it's the right layers to have and track such things.

Reply:
I guess the issue then is the definition of "dirty". We had discussed this, and we got that "a state when there is potential data loss if container is closed" could be a good one. So having the runtime as dirty while uploadBlob made sense. We discussed also having a separate runtime API like runtime.hasPendingBlobs but we decided dirty option was more convenient (single property for user to check for potential data loss state)

Reply:
If we want to push into this direction, I'd love to see the following things happened:
Saving container state would preserve these blobs (or rehydration) in all cases. In other words, when we communicate dirty, saved document state has local changes (not empty). Including cases like
- blobs uploaded in "attaching" state to be included in getPendingBlobs()
- If blobs pending upload (including above case) are saved in getPendingBlobs(), they are applied to data model on container rehydration. It's not the case today, because uploadBlob() is async API (and caller has no result to hold on to while upload is pending), so it does not matter - if we consider container is dirty in this state, then we acknowledge that we have data loss.

I have a feeling that fixing cases above forces us toward a model that follows "transition to offline" mode (from BlobManager perspective) when saving container. And if it's the case, maybe it's better to switch to fully sync model (i.e. blob uploads are synchronous from model perspective, upload block all the changes that follows).

The assumption I have here - apps (like Outlook, WB) who use dirty event to save local state have consistent view into the world.

Reply:
@dannimad, please define data loss. If I save local state of container (Container.closeAndGetPendingLocalState) and do not have these changes, is this data loss? If not (this is today's behavior) - than container should not be dirty. If it's data loss (and we indicate it so through dirty flag), then I expect data to be saved and rehydrated properly - that's what our clients would expect. But that's not possible in today's system without changing all uploads to apply to local model synchronously - either immediately, or by transitioning to "offline" mode on getPendingBlobs(). But latter does not help, as no changes yet were applied to local data, so it's still a data loss. So, we have to always operate in sync mode.

Basically, you can't change just one piece of the puzzle - the whole castle starts to collapse :) I really want to see consistent model here. It's already not very consistent (due to transitionToOffline()), but I have a hard time seeing why we need to make it even less consistent and hard to understand (mostly - from how it's projected to the users).

I think you look at it only through perspective of online container, with an app that has a layer of "lies" on top of it (like Bohemia - they have a layer that applies image to presentation layer), and container that never gets saved. I want to see at full picture (all workflows) and ensure we present consistent world through our APIs.